### ISSUE TITLE
Issues with workbenches & devFlags

### ISSUE THREAD

**Issue by BlueCog** [opendatahub-io/opendatahub-operator#780]
I'm attempting to air-gap `odh-2.4.0`. The operator is operational, and now I'm trying to air-gap the various components using devFlags.

So far, I've successfully air-gapped `dashboard` and `datasciencepipelines` without any problems:

```
    dashboard:
      devFlags:
        manifests:
          - contextDir: components/dashboard
            sourcePath: base
            uri: >-
              https://repo.com/repository/opendatahub/odh-manifests.tar.gz
    datasciencepipelines:
      devFlags:
        manifests:
          - contextDir: components/data-science-pipelines-operator
            sourcePath: base
            uri: >-
                https://repo.com/repository/opendatahub/odh-manifests.tar.gz
```
Now, I'm moving on to `workbenches`, but I'm encountering some troubles...

I've followed this function:
https://github.com/opendatahub-io/opendatahub-operator/blob/odh-2.4.0/components/workbenches/workbenches.go#L33

I've also reviewed older incidents, like: https://github.com/opendatahub-io/opendatahub-operator/issues/719, which provided more details. However, the `base` folder in `odh-notebook-controller` seems to have been removed in newer versions.

Here is the relevant part of my data science cluster YAML for `workbenches`:
```
    workbenches:
      devFlags:
        manifests:
          - contextDir: components/odh-notebook-controller           
            uri: >-
              https://repo.com/repository/opendatahub/odh-manifests.tar.gz
          - contextDir: components/notebook-controller                                    
            uri: >-
              https://repo.com/repository/opendatahub/odh-manifests.tar.gz
          - contextDir: components/notebook-images                                                    
            sourcePath: overlays/additional                                                
            uri: >-
              https:///repo.com/repository/opendatahub/notebooks-manifests.tar.gz
      managementState: Managed
```

The folder structure of `odh-manifests.tar.gz` is: `odh-manifests/components/odh-notebook-controller`

I've edited the `kustomization.yaml` files to reflect the air-gapped images. `Base` for `odh-notebook-controller` and `overlays/openshift` for `kf-notebook-controller`. Some offline testing, like `kustomize build . | grep image:`, is successful.

However, the `kustomization.yaml` is not being applied. Also, when I check the local file system at `/opt/manifests/odh-notebook-controller/` of the operator, the `kustomization.yaml` is the default one, not the custom one I've edited and hosted at `odh-manifests.tar.gz`.

I've tried multiple scenarios:

- with or without `sourcePath`
- Different `contextDir` (but the function seems specific to this).
- Restarting the operator (to flush any local data).
- Removing the components folder from the `tar.gz` (which seems not to matter for the function).
- Editing all the `kustomization` files and deployments.
...

At one point, I was successful in editing `odh-notebook-controller`, but I never had a working scenario for `kf-notebook-controller`.

Did I miss something?

_P.S. It would be helpful if the operator logged some failures or issues with devFlags._

**Comment by zdtsw** [comment]
I guess the whole reason with using devFlags is you want to run ODH in air-gapped env in your orgnization.
but not because you want to customize certain filed in the mainfests for components.

this is a very interesting use case.
we do have some community users want to run ODH in air-gapped see https://github.com/opendatahub-io/opendatahub-operator/issues?q=label%3Adisconnected but i do not think we are officially support this. 
But I assume your  hack approach can be a way out, if you are able to get all the components mainfests tarball and all images behind your proxy. 

from what you described above, if dashboard and dsp are working then i think the problem does not lie in the networking or permission part.
I am sure we do have error logging enabled if it is not able to download tarballs.
Could you try this:

- have a clean default DSC CR wihtout devFlags for workbench
- restart operator pod first
- disable workbenches in DSC CR
- login to operator pod and remove the workbenches folders
- add devFlags in the DSC CR for workbenches and enable workbench
- login to operator pod and see if your customized tarballs are decompressed there.
- if above steps are all ok, the last one is to do  a `grep` on image under such folders and see if they are pointing to your customied images.

**Comment by shalberd** [comment]
> I guess the whole reason with using devFlags is you want to run ODH in air-gapped env in your organization.
but not because you want to customize certain filed in the mainfests for components.

yes for him and for me. That is the sole reason. We mirrored the images to a local repo (in his case, a repo without authentication. i.e. docker pull secret imagePullSecrets component serviceaccount mods in the manifests needed), and adapted the revelant part 

- in the images section of kustomization.yaml or. 
- for imagestreams, in the tag.docker.from.name part.

@BlueCog was able to delete the operator pod, pod restarted, still same behavior.
We tried recursively deleting /opt/manifests directory to get a clean slate, which showed permission denied. User is cluster-admin. Which is strange, I know this worked for v1 operator version. Permission settings wrong or different since operator v2 maybe?
Also, deleting the operator pod (we did that) should lead to /opt/manifests being populated freshly, too, cause there is no PVC / volume behind that location. But, even after having done that, still, the default state of the manifests for kf-notebook-controller is there.

See also Slack https://odh-io.slack.com/archives/C0330L52N22/p1702197339770199?thread_ts=1702069081.999439&cid=C0330L52N22

**Comment by zdtsw** [comment]
what i did is just replace the tag to "latest" and push to my forked kubeflow repo on v1.7-branch
then update DSC to 
![Screenshot from 2023-12-11 17-19-09](https://github.com/opendatahub-io/opendatahub-operator/assets/915053/b5d0386d-831b-4396-bcd9-45d2a06c1b1f)
then check screenshot from my operator pod
![Screenshot from 2023-12-11 17-18-31](https://github.com/opendatahub-io/opendatahub-operator/assets/915053/fd2d1ccc-c293-471a-a6ae-d0bcd2a28099)


looks all good, but
deployment did not get updates => pod did not restart with new "latest" image
which is ODD....

**Comment by AjayJagan** [comment]
@zdtsw @BlueCog -> If you feel this is okay to move to JIRA and track the issue there, please let me know I can create the JIRA for the same. Also, please feel free to create the JIRA if you want to :)

**Comment by BlueCog** [comment]
Hello @AjayJagan I took a shot at creating the JIRA issue. I'm not completely familiar with your JIRA process/metadata, etc., so feel free to correct where needed. ;)

https://issues.redhat.com/browse/RHOAIENG-988

P.s. for anyone who follows this proces of air gapping. For now I was able to air gap the operator and manifests by building the entire operator and providing my own manifest definitions during the build proces. Feel free to contact me if you have any questions about that (my wish is to create a blog post about this proces in the near future)
